### DISCUSSION TITLE
Accessing vector values from user defined types

### DISCUSSION THREAD

**Discussion by abarun22** [idaholab/moose#24781] [Q&A General]
I was looking at the following functionality and trying to understand how a vector defined with a userdefined type such as "RealVectorValue" can be accessed with an unconventional paranthesis just after the vector object. 
```
RealVectorValue
ParsedVectorAux::computeValue()
{
  RealVectorValue value;
  for (const auto i : make_range(Moose::dim))
  {
    for (const auto j : make_range(_nargs))
      _func_params[j] = (*_args[j])[_qp];

    for (const auto j : make_range(_n_vector_args))
      _func_params[_nargs + j] = (*_vector_args[j])[_qp](i);

    if (_use_xyzt)
    {
      for (const auto j : make_range(Moose::dim))
        _func_params[_nargs + _n_vector_args + j] =
            isNodal() ? (*_current_node)(j) : _q_point[_qp](j);
      _func_params[_nargs + _n_vector_args + 3] = _t;
    }
    value(i) = evaluate(_func_F[i]);
  }
  return value;
}
```
I see that RealVectorValue is defined in "VectorValue" class and that by saying `vector(i)` i think we are calling the constructor of this class. I am unclear how the elements of the vector themselves are stored, as "vector_value.h" does not seem to provide any clue. Could someone explain how this really works, especially the reason behind the usage of a parantheseis to access these vector elements?
Kind regards,
Arun

**Comment by GiudGiud** [comment]
Hello

This is because we have defined the parenthesis operator on these classes ourselves. You can override the standard operators in C++, +, - etc

see the 6th or so public function definiton on this page:
https://mooseframework.inl.gov/docs/doxygen/libmesh/classlibMesh_1_1VectorValue.html

It does the () operator

Guillaume

**Reply by abarun22** [reply]
Thanks a lot Guillaume. You cleared me a long time, wanting to ask question.
Kind regards,
Arun
